packages/gatsby/src/redux/nodes.ts

```diff
@@ -452,75 +452,32 @@ function addNodeToBucketWithElemMatch(
   }
 }
 
-const binarySearchAsc = (
-  values: Array<FilterValue>, // Assume ordered asc
-  needle: FilterValue
-): [number, number] | undefined => {
+// Returns the first index in `values` where `pred` is false, given
+// that `needle` returns true for some, possibly empty, prefix of
+// `values` and returns `true` for the remaining, possibly empty,
+// portion of `values`. If all `values` match, returns
+// `values.length`. If no `values` match, returns `0`, the first index
+// where `needle` is false.
+function binarySearch<T>(values: Array<T>, pred: (cur: T) => boolean): number {
```

Comment:
It's possible we want to leave these specialized for `lte` and `gte` instead of using this predicate approach. Happy to change it.